File: April 22/Next element with greater frequency gfg.cpp
class Solution{
public:
    vector<int> print_next_greater_freq(int arr[],int n)
    {
        map<int,int> count;
        for(int i=0;i<n;i++) count[arr[i]]++;
        
        stack<int> st;
        st.push(arr[n-1]);
        
        vector<int> ans(n,-1);
        for(int i=n-2;i>=0;i--){
            while(!st.empty()){
                if(count[st.top()] > count[arr[i]]){
                    ans[i]=st.top();
                    break;
                }
                st.pop();
            }
            st.push(arr[i]);
        }
        return ans;
    }
};

File: JULY 2025/08 July/Next element with greater frequency gfg.java
class Solution {
    public ArrayList<Integer> findGreater(int[] arr) {
        // code here
        Map<Integer,Integer> freq = new HashMap<>();
        for(int num:arr)
            freq.put(num, freq.getOrDefault(num,0)+1);
        Stack<Integer> stk = new Stack<>();
        int N = arr.length;
        ArrayList<Integer> ans = new ArrayList<>();
        for(int i=N-1; i>=0; i--)
        {
            while(stk.size()>0 && freq.get(stk.peek())<=freq.get(arr[i]))
                stk.pop();
            
            if(stk.size()==0)
                ans.add(-1);
            else
                ans.add(stk.peek());
            stk.push(arr[i]);
        }
        Collections.reverse(ans);
        return ans;
    }
}
